File: solutions/143.reorder-list.py
#
# @lc app=leetcode id=143 lang=python3
#
# [143] Reorder List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if not head or not head.next or not head.next.next:
            return
        
        slow = head
        fast = head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next       # two steps

        cut_middle = slow
        slow = slow.next
        cut_middle.next = None      # This is the end of the new list

        tail = None                 # Reverse ordered second part of the list
        while slow:
            temp = slow.next
            slow.next = tail
            tail = slow
            slow = temp

        while head and tail:
            temp_head = head.next
            head.next = tail
            temp_tail = tail.next
            tail.next = temp_head
            head = head.next.next
            tail = temp_tail

# @lc code=end

File: solutions/143.reorder-list.java
/*
 * @lc app=leetcode id=143 lang=java
 *
 * [143] Reorder List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        if (head == null || head.next == null || head.next.next == null) {
            return;
        }

        ListNode slow = head;   // it will point to the middle once fast reach the end
        ListNode fast = head.next;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;   // do two steps
        }

        ListNode cutMiddle = slow;
        slow = slow.next;
        cutMiddle.next = null;    // There will be the end of the reordered list

        ListNode tail = null;   // it will contain second part of list in the reverse order
        while (slow != null) {
            ListNode temp = slow.next;
            slow.next = tail;
            tail = slow;
            slow = temp;
        }

        while (head != null && tail != null) {
            ListNode tempHead = head.next;
            head.next = tail;
            ListNode tempTail = tail.next;
            tail.next = tempHead;
            tail = tempTail;
            head = head.next.next;
        }
    }
}
// @lc code=end

